test/util.sh

```diff
@@ -0,0 +1,203 @@
+function os::test::expect_code_and_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}"
+}
+
+# expect_code_and_not_text runs the cmd and expects the given exit code
+# as well as running a grep test to ensure the given string is not in the output
+function os::test::expect_code_and_not_text() {
+	cmd=$1
+	expected_cmd_code=$2
+	expected_text=$3
+
+	os::test::expect_exit_code_run_grep "${cmd}" "specific_code_func ${expected_cmd_code}" "${expected_text}" "failure_func"
+}
+
+# TEST EXECUTION FUNCTION
+
+# expect_exit_code_and_text runs the provided test command and expects a specific exit code from that
+# command as well as the success of a specified `grep` invocation. Output from the command to be 
+# tested is suppressed unless either `VERBOSE=1` or the test fails.
+function os::test::expect_exit_code_run_grep() {
+	cmd=$1
+	# default expected cmd code to 0 for success
+	cmd_eval_func=${2:-success_func}
+	# default grep args to a something that always passes, even for zero-length input
+	grep_args=${3:-'$'} 
+	# default expected test code to 0 for success
+	test_eval_func=${4:-success_func}
+
+	echo "Running  ${cmd}..."
+	
+	set +o errexit
+	cmd_output=$(${cmd} 2>&1)
+	cmd_result=$?
+	cmd_succeeded=$( ${cmd_eval_func} "${cmd_result}"; echo $? )
+
+	test_output=$(echo "${cmd_output}" | grep -E "${grep_args}" 2>&1)
+	test_result=$?
+	test_succeeded=$( ${test_eval_func} "${test_result}"; echo $? )
+	set -o errexit
+
+	if (( $cmd_succeeded && $test_succeeded )); then
+		overwrite_green "SUCCESS: ${cmd}"
+
+		if [[ -n ${VERBOSE-} ]]; then
+			print_results "${cmd_output}"
+		fi
+
+		return 0
+	else
+		cause=$(assemble_causes "${cmd_succeeded}" "${test_succeeded}")
+		
+		overwrite_red "FAILURE: ${cmd}: ${cause}"
+		print_results "${cmd_output}"
+
+		return 1
+	fi
+}
+
+# success_func determines if the input exit code denotes success
+function success_func() {
+	exit_code=$1
+
+	# use a negated test to get output correct for (( ))
+	[[ "${exit_code}" -ne "0" ]]
+	return $?
+}
+
+# failure_func determines if the input exit code denotes failure
+function failure_func() {
+	exit_code=$1
+
+	# use a negated test to get output correct for (( ))
+	[[ "${exit_code}" -eq "0" ]]
+	return $?
+}
+
+# specific_code_func determines if the input exit code matches the given code
+function specific_code_func() {
+	expected_code=$1
+	exit_code=$2
+
+	# use a negated test to get output correct for (( ))
+	[[ "${exit_code}" -ne "${expected_code}" ]]
+	return $?
+}
+
+# colors for test output
+readonly      reset=$(tput sgr0)
+readonly        red=$(tput setaf 1)
+readonly      green=$(tput setaf 2)
+readonly clear_last=$(tput cuu 1 && tput el)
+
+# overwrite_red overwrites the last line with the given text in green
+function overwrite_green() {
+	echo ${clear_last}${green}${@}${reset}
```

Comment:
I thought about this last night and I think `echo -n` would work. This really falls under that category of things you really enjoy labeling as "not important enough to spend time on". After I noticed it was broken, since my intent was *not* to build a nice color library, I gave up spending my time on it because it was such an inconsequential thing. If you really feel *this* strongly about it, I will write us a nice function suite for text manipulation.